agent/engine/execcmd/agent_version.go

```diff
@@ -15,68 +15,107 @@ package execcmd
-// The version must have only integers representing major, minor, patch, and reserved; delimited by dots (e.g. "1.2.3.4").
-// An error is returned if any of the versions passed as parameter is malformed.
-// example:
-// compareAgentVersion("3.0.236.0", "3.0.237.0") returns -1, nil
-// compareAgentVersion("3.0.236.0", "3.0.236.0") returns 0, nil
-// compareAgentVersion("3.0.237.0", "3.0.236.0") returns 1, nil
-// compareAgentVersion("a.b.c.d", "3.0.236.0") returns -999, error
-func compareAgentVersion(v1Str, v2Str string) (int, error) {
-	v1Arr := strings.Split(v1Str, versionDelimiter)
-	v2Arr := strings.Split(v2Str, versionDelimiter)
+type agentVersion []uint
 
-	if len(v1Arr) != len(v2Arr) {
-		return badInputReturn, errors.New(badFormatErrorStr)
+// compare compares this agentVersion against another one passed as parameter.
+// returns -1 if this < av2, 0 this == av2, and 1 if this > av2.
+// example:
+// [3, 0, 236, 0].compare([3, 0, 237, 0]) returns -1
+// [3, 0, 236, 0].compare([3, 0, 236, 0]) returns 0
+// [3, 0, 237, 0].compare([3, 0, 236, 0]) returns 1
+// [3].compare([2, 0, 236])               returns 1
+func (av agentVersion) compare(av2 agentVersion) int {
+	av1Len := len(av)
+	av2Len := len(av2)
+	maxLen := av1Len
+	if av2Len > maxLen {
+		maxLen = av2Len
 	}
-	for i := 0; i < len(v1Arr); i++ {
-		v1Int, err := strconv.ParseUint(v1Arr[i], 10, 32)
-		if err != nil {
-			return badInputReturn, errors.New(badFormatErrorStr)
+	for i := 0; i < maxLen; i++ {
+		v1 := uint(0)
+		v2 := uint(0)
+		if i < av1Len {
+			v1 = av[i]
 		}
-		v2Int, err := strconv.ParseUint(v2Arr[i], 10, 32)
-		if err != nil {
-			return badInputReturn, errors.New(badFormatErrorStr)
+		if i < av2Len {
+			v2 = av2[i]
 		}
-
-		if v1Int < v2Int {
-			return -1, nil
-		} else if v1Int > v2Int {
-			return 1, nil
+		if v1 < v2 {
+			return -1
+		} else if v1 > v2 {
+			return 1
 		}
 	}
-	return 0, nil
+	return 0
 }
 
-func determineLatestVersion(versions []string) (string, error) {
-	if len(versions) == 0 {
-		return "", errors.New("no versions to compare were provided")
+// String the string representation of an agentVersion
+// e.g. [3, 0, 236, 0] -> "3.0.236.0"
+func (av agentVersion) String() string {
+	var avStr []string
+	for _, e := range av {
+		avStr = append(avStr, strconv.Itoa(int(e)))
 	}
-	latest := "0.0.0.0"
-	for _, v := range versions {
-		comp, err := compareAgentVersion(v, latest)
+	return strings.Join(avStr, versionDelimiter)
+}
+
+func parseAgentVersion(version string) (agentVersion, error) {
+	var parsedVersion []uint
+	vArr := strings.Split(version, versionDelimiter)
+	for _, part := range vArr {
+		pInt, err := strconv.ParseUint(part, 10, 32)
 		if err != nil {
-			return "", err
+			return nil, errors.New(badFormatErrorStr)
+		}
+		parsedVersion = append(parsedVersion, uint(pInt))
+	}
+	return parsedVersion, nil
+}
+
+type agentVersionList []agentVersion
```

Comment:
nit: why not just use `[]agentVersion` instead? defining this as a new type does not bring too much benefit and makes it a bit harder to read (cannot directly tell what data type it actually is)

Reply:
The reason I created a new type is that I wanted to make `[]agentVersion` sortable. I will check if I can still do that just by using `[]agentVersion` directly... If not, I think it's still a good recommendation to change this to make it clear it's just for sorting purposes

Reply:
Alright, I renamed `agentVersionList` to `byAgentVersion` and I only use this type for sorting, the rest of the functions use `[]agentVersion` directly. And you're right that this is more readable :)

Reply:
good point, thanks for the explanation! I thought `sort.Sort()` requires implementing comparator on the type that needs to be compared (i.e. `agentVersion`), didn't realize that it should be implemented on the slice type too

meantime, just curious, would it work to directly implement the interface methods on `[]agentVersion`? this way we won't need the added type `byAgentVersion` either

```go
func (bav []agentVersion) Len() int {
	return len(bav)
}
```